### ISSUE TITLE
Git lfs files skipped

### ISSUE THREAD

**Issue by nusasara** [git-lfs/git-lfs#3008]
Hello!

We are planning to convert repo to LFS repo with BFG. This worked well on test server. But we also tried backup/restore process if something goes wrong. What is bothering us is  this. We push some files to LFS and than restore repo. When we push some LFS file again (the file it was pushed before) we get this at push:   Git LFS: (0 of 0 files, 1 skipped)

I understand that this means that the file is already on the server. Is there any way to remove them before push, to have clean state and push the files every time. Where are lfs files stored, how can we delete them? I was looking in location that is in configuration gitlab_rails['lfs_storage_path'] but there are some HEX names of folders. How do I know which file is for which project?

If we will try a lot of times we will have a lot of big files that are not used. 

BR Denis

**Comment by nusasara** [comment]
Thank you for all the info. Will try git lfs push -all and will try git lfs migrate. About location of lfs files, I will try my luck on gitlab forum. Can I contact you on email or here if I need some assistance with git lfs migrate. Thank you again. BR Denis

**Comment by ttaylorr** [comment]
> Thank you for all the info. Will try git lfs push -all and will try git lfs migrate. About location of lfs files, I will try my luck on gitlab forum.

Sounds great to me, thanks!

> Can I contact you on email or here if I need some assistance with git lfs migrate.

Here is preferred, since it can be helpful for others as well.

**Comment by nusasara** [comment]
Hi,

As you recommend that best practice for migration is by using the “git lfs migrate” tool.

We are migrating a huge repo 50+gb to lfs. The issue here is that we have a lot of image files stored in the repo and we would like to migrate them. I know it isn't good practice for git and 50GB, but we have what we have and we need solution that only build server and build users checkout these large files and not every developers (they won't have lfs enabled and will checkout only links to files) and checkout speed will increase by 100x.

Our scenario is the following:
-	Repo size 50+gb
-	9000+ commits
-	70+ branches
-	2000+ merge requests
-	We want to migrate all the *.gz, *.xz & *.qcow2 files and add them track them with lfs

The goal we want to achieve is to migrate all those files to lfs and that’s in the way that we do not lose the structure and its history. So could you please provide us some hints/steps/examples what should be a best practice, in this case?

BR Denis

**Comment by ttaylorr** [comment]
> * We want to migrate all the *.gz, *.xz & *.qcow2 files and add them track them with lfs

The following should work:

```bash
git lfs migrate import --everything --include="*.gz:*.xz:*.gcow"
git push --force --all origin
```

With the caveat that everyone will have to `fetch` the updated history and reset their local branches as appropriate.

**Comment by nusasara** [comment]
We tried this. Had to use:
git lfs migrate import --everything --include="*.gz,*.xz,*.gcow"
instead of
git lfs migrate import --everything --include="*.gz:*.xz:*.gcow"

It makes lfs-object files, but repo size stays the same on server so we use double space on disk. And if we checkout with no lfs enabled(as it is planned for developers) we don't get just 1GB of repo files, we get the whole 33GB al files. Any thought? What can we try?

BR Denis

**Comment by ttaylorr** [comment]
@nusasara It sounds like the migration is working as expected, but leaving extra Git objects on the server. You should run `git-prune` or `git-gc` to remove these unreferenced objects.

**Comment by nusasara** [comment]
Thank you for your answer. I found this also when google-ing. Where do I have to do this? Do a git clone to one folder and than do git-prune or git-gx or is it on server side. I will take a look at the documentation.

**Comment by ttaylorr** [comment]
> Where do I have to do this? Do a git clone to one folder and than do git-prune or git-gx or is it on server side.

In a shell on your computer within the repository should be sufficient.

**Comment by somsakpanom** [comment]
ผมรับค่าตอบแทนไหมคับ

ในวันที่ อ. 29 พ.ค. 2018 07:19 Taylor Blau <notifications@github.com>
เขียนว่า:

**Comment by nusasara** [comment]
We tried that but didn't help. How is this command gonna help on server side, if we do it just on client side? Do I have to do git push after that? But size on client size didn't drop eighter. And after migration to lfs, repo size is still the same, it made a 33GB (the previous size of a repo) large file in objects folder, named pack-41495bb0377ea13b448c261d7176ad27829a44d2.pack. And also when we tired to pull files without having lfs installed on client side it cloned 33GB. When we tried with BFG the simptoms were the same (same size of repo on server, double the size of LFS files) but when cloning without lfs installed it cloned just 1GB which is corrent. Any ideas?

BR Denis

**Comment by ttaylorr** [comment]
> We tried that but didn't help. How is this command gonna help on server side, if we do it just on client side?

I misunderstood your goal. It is up to the server to delete Git objects that are orphaned/stale/etc, and it isn't something that the client has absolute control over.

> And after migration to lfs, repo size is still the same, it made a 33GB (the previous size of a repo) large file in objects folder, named pack-41495bb0377ea13b448c261d7176ad27829a44d2.pack.

Not sure what's going on here. You could try some incantations of `git-repack(1)` (https://git-scm.com/docs/git-repack) to help with this.

Do you have a reproducible example of this problem?

**Comment by nusasara** [comment]
Hello!

Thx for every advice. After some playing around we managed to make it work. We tested it on test environment and everything was ok. Than we tried the same thing on production and everything was ok. But when we tried to do it for real on production we had a problem. 
I know this is probably a gitlab specific problem, but maybe you colud help as I am hitting dead end. 
New project was created on production environemnt. I forgot to enable lfs and when pushing it didn't work. But now for some when I do git lfs push for any project when adding some LFS file (it can be small so it isn't some timeout issue) it keeps pushing, everytime saying "LFS: Access forbidden. Check your access level." trying again and every time it tries for another time it makes a new folder in cache folder of lfs-objects. I know that project has LFS enabled, because if I disable it it failes at:
trace git-lfs: HTTP: POST .../info/lfs/locks/verify
trace git-lfs: HTTP: POST .../info/lfs/objects/batch

And when LFS is enabled it goes pass this but fails here:

22:48:00.649880 trace git-lfs: HTTP: {"objects":[{"oid":"7571a85db75aa74d7902aec1fd52d57d2f61bd48d477d319f05e7cbba46155ce","size":2088566784,"actions":{"upload":{"href":"../gitlab-lfs/objects/7571a85db75aa74d7902aec1fd52d57d2f61bd48d477d319f05e7cbba46155ce/2088566784","header":{"Authorization":"Basic ZGVuaXNvOiE0S3JhdG5vU3Jhbmpldkl0YWxpamkh"}}}}]}
22:48:00.649880 trace git-lfs: tq: starting transfer adapter "basic"
Git LFS: (0 of 1 files) 0 B / 1.95 GB                                          22:48:00.650380 trace git-lfs: HTTP: PUT ../gitlab-lfs/objects/7571a85db75aa74d7902aec1fd52d57d2f61bd48d477d319f05e7cbba46155ce/2088566784
Git LFS: (0 of 1 files) 1.95 GB / 1.95 GB                                      22:50:48.049854 trace git-lfs: HTTP: 403
22:50:48.049854 trace git-lfs: HTTP: {"message":"Access forbidden. Check your access level.","documentation_url":"https://git.comtrade.com/help"}
22:50:48.050354 trace git-lfs: tq: retrying object 7571a85db75aa74d7902aec1fd52d57d2f61bd48d477d319f05e7cbba46155ce: LFS: Access forbidden. Check your access level.
22:50:48.050354 trace git-lfs: tq: enqueue retry #1 for "7571a85db75aa74d7902aec1fd52d57d2f61bd48d477d319f05e7cbba46155ce" (size: 2088566784)

Any idea what could be the problem. Will write to gitlab forum, butt we have CE version and until now none of the posts were resolved.

Why I think it is a gitlab problem, as this still works on test environment. I'm looking for differences but can't find them.

BR Denis

**Comment by ttaylorr** [comment]
> Any idea what could be the problem. Will write to gitlab forum, butt we have CE version and until now none of the posts were resolved.

I agree that this seems like a problem pertinent to the way that your instance of GitLab is operating. From the trace that you posted above, we know that Git LFS is sending credentials to GitLab, but GitLab is rejecting them. Unless your credential helper or `~/.netrc` are misconfigured and sending the wrong credentials, I think that the GitLab forum will have the best idea what's going on.
